explanation/performance/perf-pgo.md

```diff
@@ -0,0 +1,153 @@
+[31] .debug_abbrev        PROGBITS     0000000000000000 001bcbf3 000110c4  0        0   0  1
+[...]
+```
+
+If the `Flags` field has `C` in it, this means that `dwz` was used.  You will then need to figure out how to disable its execution during the build.
+
+### Using `perf` to obtain profile data
+
+With OpenSSL prepared and ready to be profiled, it's now time to use `perf` to obtain profiling data from our workload.  If you are on an Intel system, you can invoke `perf` using:
+
+```bash
+$ sudo perf record \
+	-e br_inst_retired.near_taken \
+	--branch-any \
+	--all-cpus \
+	--output openssl-nonpgo.perfdata \
+	-- openssl speed -mr -seconds 60 -evp md5 sha512 rsa2048 aes-256-cbc
+```
+
+On an AMD system, use the following invocation:
+
+```bash
+$ sudo perf record \
+	-e 'cpu/event=0xc4,umask=0x0,name=ex_ret_brn_tkn/' \
+	--branch-any \
+	--all-cpus \
+	--output openssl-nonpgo.perfdata \
+	-- openssl speed -mr -seconds 60 -evp md5 sha512 rsa2048 aes-256-cbc
+```
+
+Note that the only thing that differs is the PMU event to be monitored.
+
+After the command has finished running, you should see a file named `openssl-nonpgo.perfdata` in your current directory.  This is the file we will now need to convert to `gcov`.
+
+### Converting `perfdata` to `gcov` with `autofdo`
+
+With `autofdo` installed, you can convert the `perfdata` generated in the last step to `gcov` by doing:
+
+```bash
+$ sudo chown ubuntu:ubuntu openssl-nonpgo.perfdata
+$ create_gcov \
+	--binary /usr/bin/openssl \
+	--gcov openssl-nonpgo.gcov \
+	--profile openssl-nonpgo.perfdata \
+	--gcov_version 2 \
+	-use_lbr false
+```
+
+`create_gcov` is verbose and will display several messages that may look like something is wrong; you should inspect its output and also make sure to check the `$?` shell variable to make sure that it has finished successfully.  You will then have a file called `openssl-nonpgo.gcov` in your current directory, which is the `gcov` file we will use to feed back to GCC when recompiling OpenSSL.  You can inspect the file by using the `dump_gcov` tool.
+
+### Rebuilding OpenSSL with PGO
+
+We have everything we need to rebuild OpenSSL and make use of our profile data.  The most important thing now is to set the `CFLAGS` environment variable properly so that GCC can find and use the profile we generated in the previous step.
+
+How you perform this `CFLAGS` adjustment depends on how you built the software in the first place, so we won't cover this part here.  The resulting `CFLAGS` variable should have the following GCC option, though:
+
+```
+-fauto-profile=/path/to/openssl-nonpgo.gcov
+```
+
+Make sure to adjust the path to the `openssl-nonpgo.gcov` file.
+
+Also note that enabling PGO will make GCC automatically enable several optimization passes that are usually disable, even when using `-O2`.  This can lead to new warnings/errors in the code, which might require either the code to be adjusted or the use of extra compilation flags to suppress these warnings.
```

Comment:
```suggestion
Also note that enabling PGO will make GCC automatically enable several optimization passes that are usually disabled, even when using `-O2`.  This can lead to new warnings/errors in the code, which might require either the code to be adjusted or the use of extra compilation flags to suppress these warnings.
```